### DISCUSSION TITLE
context window limitation.

### DISCUSSION THREAD

**Discussion by starthlo** [gptme/gptme#138] [General]
Hi, I tried to run `/shell ls -R` command to read all directories. It's token usage was 289628 that exceeds `claude-3.5-sonnect` model's context window. I reviewed your `reduce.py` and it truncated message codeblocks to the first and last `X` lines, keeping the rest as `[...]`.
I think we should keep all commands  and codeblocks while `reduce` processing. Actually, my shell command was essential for achieving my goals using `gptme`.  

![Screenshot from 2024-09-24 12-15-35](https://github.com/user-attachments/assets/550baf71-c397-4c04-8777-5fa1822aa0ea)

To address this problem, I attempted to store that codeblock on vector database and reuse it in conversation using the RAG approach. However, this was not a viable solution because we have already reached the maximum token limit.

How can we solve this kind of issue?

**Comment by ErikBjare** [comment]
If you really want to use the output of such a command, you should store it to file and use grep or some RAG tool to retrieve what you want.

You can temporarily fix your conversation by removing the massive line in the `conversation.jsonl`. I should probably make the recovery from such a situation less in need of manual intervention.

But looking at the command, you seem to be dumping an entire node_modules into context. Instead you could use something like `ls $(git ls-files)` (assuming your repo isn't huge).

I recently changed the behavior you describe with lines, it now takes the first 2000 + last 8000 tokens instead: https://github.com/ErikBjare/gptme/commit/8a62859b93005663f0150ea4d7030c0bd92a382d

RAG in general is tracked in #59.

I was recently working on a local RAG thing in #135, but not sure if it'll get merged anytime soon.

**Comment by starthlo** [comment]
I understand, but I need to dump the entire code block or output in the context without any truncation. Initially, I thought I could achieve this using RAG. However, I've realized it's not feasible to store it in a vector database and retrieve it later using RAG, as we have already reached the model's maximum context window size. Do you know the best solutions to handle this kind of situation seamlessly?

**Comment by TimeToBuildBob** [comment]
Coming back to this in 2026 — a few things have improved since this was posted that help with this class of problem:

**Use targeted queries instead of dumping everything**

`ls -R` on a project with `node_modules` will overflow any context window (yours was 289K tokens). The right pattern is to narrow what you need:

```bash
# List only tracked files (excludes node_modules, build artifacts)
git ls-files

# Find specific patterns
find . -name "*.py" -not -path "*/node_modules/*"

# Search inside files
grep -r "function_name" --include="*.js" -l
```

**Store large output to file, then search**

Instead of keeping it in context, store to a file and use gptme's `read` tool with line ranges, or grep/rg for the specific part you need:

```bash
ls -R > /tmp/file-list.txt
grep "something-specific" /tmp/file-list.txt
```

**Models now have much larger context windows**

Since 2024, most frontier models have 128K–200K token contexts (Claude claude-3.5 Sonnet: 200K, Gemini 1.5 Pro: 1M), which handles most real codebases. If your project genuinely requires >200K of file listing, the problem is usually that you are including generated/dependency files that do not need to be in context.

**The `read` tool supports partial reads**

gptme's `read` tool accepts `head`, `tail`, and `offset` parameters so you can paginate through large files without loading them all at once.

The core insight: LLMs work better with targeted context than with everything-at-once. The "dump the whole tree" approach is rarely what you actually need — usually you need "find where X is defined" or "list all files in module Y", which targeted queries can answer without hitting context limits.
